core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -567,15 +632,17 @@ class ReplicaManager(val config: KafkaConfig,
         } catch {
           // NOTE: Failed produce requests metric is not incremented for known exceptions
           // it is supposed to indicate un-expected failures of a broker in handling a produce request
-          case e: KafkaStorageException =>
-            fatal("Halting due to unrecoverable I/O error while handling produce request: ", e)
-            Exit.halt(1)
-            (topicPartition, null)
+          case e: IOException =>
+            val dirOpt = getLogDir(topicPartition)
+            error(s"Error while appending records to partition $topicPartition in dir $dirOpt", e)
+            dirOpt.foreach(maybeAddLogFailureEvent)
+            (topicPartition, LogAppendResult(LogAppendInfo.UnknownLogAppendInfo, Some(new KafkaStorageException(e))))
```

Comment:
No.. Currently I didn't catch/try IOException for methods like Log.append(), LogSegment.append() or OffsetIndex.append(). I feel that Kafka' code will be cleaner and more reliable if we can catch IOException in the methods which handle request and generate response. The only reason not all IOException are caught in ReplicaManager or KafkaApi is that we have checkpoint file operation, ReplicaFetcherThread or LogCleaner which does IO operation without requiring external request.